include/nbl/builtin/hlsl/workgroup/shared_scan.hlsl

```diff
@@ -0,0 +1,184 @@
+    uint scanStoreIndex;
+
+    static WorkgroupScanHead create(bool isScan, T identity, uint itemCount)
+    {
+        WorkgroupScanHead<T, SubgroupOp, SharedAccessor> wsh;
+        wsh.isScan = isScan;
+        wsh.identity = identity;
+        wsh.itemCount = itemCount;
+        wsh.lastInvocation = itemCount - 1u;
+        return wsh;
+    }
+
+    T operator()(T value)
+    {
+        const uint subgroupMask = glsl::gl_SubgroupSize() - 1u;
+        SharedAccessor sharedAccessor;
+        lastInvocationInLevel = lastInvocation;
+        
+        SubgroupOp subgroupOp;
+        firstLevelScan = subgroupOp(value);
+        T scan = firstLevelScan;
+        
+        const bool isLastSubgroupInvocation = glsl::gl_SubgroupInvocationID() == glsl::gl_SubgroupSize() - 1u;
+        
+        // Since we are scanning the RESULT of the initial scan (which paired one input per subgroup invocation) 
+        // every group of 64 invocations has been coallesced into 1 result value. This means that the results of 
+        // the first SubgroupSz^2 invocations will be processed by the first subgroup and so on.
+        // Consequently, those first SubgroupSz^2 invocations will store their results on SubgroupSz scratch slots 
+        // with halfSubgroupSz padding and the next level will follow the same + the previous as an `offset`.
+        const uint offset = (gl_LocalInvocationIndex >> glsl::gl_SubgroupSizeLog2()) & ~subgroupMask; // For subgroupSz = 64, first 4095 invocations get offset 0, 4096-8191 offset 64, then 128 etc.
+        const uint memBegin = offset;
+        uint nextLevelStoreIndex = memBegin + glsl::gl_SubgroupID();
+        scanStoreIndex = gl_LocalInvocationIndex;
+        
+        bool participate = gl_LocalInvocationIndex <= lastInvocationInLevel;
+        while(lastInvocationInLevel >= glsl::gl_SubgroupSize() * glsl::gl_SubgroupSize())
+        {
+            sharedAccessor.main.workgroupExecutionAndMemoryBarrier();
+            if(participate)
+            {
+                if (any(bool2(gl_LocalInvocationIndex == lastInvocationInLevel, isLastSubgroupInvocation)))
+                { // only invocations that have the final value of the subgroupOp (inclusive scan) store their results
+                    sharedAccessor.main.set(nextLevelStoreIndex, scan);
+                }
+            }
+            sharedAccessor.main.workgroupExecutionAndMemoryBarrier();
+            participate = gl_LocalInvocationIndex <= (lastInvocationInLevel >>= glsl::gl_SubgroupSizeLog2());
+            if(participate)
+            {
+                const uint prevLevelScan = sharedAccessor.main.get(gl_LocalInvocationIndex);
+                scan = subgroupOp(prevLevelScan);
+                if(isScan)
+                {
+                    sharedAccessor.main.set(scanStoreIndex, scan);
+                }
+            }
+            if(isScan)
+            {
+                scanStoreIndex += lastInvocationInLevel + 1u;
+            }
```

Comment:
assuming implementation of (https://github.com/Devsh-Graphics-Programming/Nabla/pull/519/files#r1327215310) this should be
```cpp
if (participate)
{
   if (subgroupElect())
      accessor.set(scanStoreIndex,scan);
   if (isScan)
      scanStoreIndex += lastInvocationInLevel+1;
   participate = gl_LocalInvocationIndex <= (lastInvocationInLevel >>= glsl::gl_SubgroupSizeLog2());
}
accessor.workgroupExecutionAndMemoryBarrier();
if (participate)
{
   scan = accessor.get(scanLoadIndex);
   if (isScan)
      scanLoadIndex += lastInvocationInLevel+1;
}
accessor.workgroupExecutionAndMemoryBarrier();
```

Reply:
btw note that both indices get incremented by the same value

Reply:
ok one more note, if you actually remove the shared mem size optimization for a reduction, you can remove the barrier between get and set
```cpp
bool participate = gl_LocalInvocationIndex <= lastInvocationInLevel;
while(lastInvocationInLevel >= glsl::gl_SubgroupSize() * glsl::gl_SubgroupSize())
{
if (participate)
{
   if (subgroupElect())
      accessor.set(scanLoadIndex-loadStoreIndexDiff,scan);
   participate = gl_LocalInvocationIndex <= (lastInvocationInLevel >>= glsl::gl_SubgroupSizeLog2());
}
accessor.workgroupExecutionAndMemoryBarrier();
if (participate)
{
   const uint prevLevelScan = accessor.get(scanLoadIndex);
   scan = subgroupOp(prevLevelScan);
   scanLoadIndex += lastInvocationInLevel+1;
}
}
```

Reply:
btw the only reason to keep `scanLoadIndex += lastInvocationInLevel+1;` outside an `if (participate)` is in order to pass it to Tail and avoid convoluted logic to get it back